### ISSUE TITLE
x/tools/godoc: Dockerize and convert to run on GAE Flex

### ISSUE THREAD

**Issue by andybons** [golang/go#27205]
godoc relies on https://golang.org/doc/go1.10#goroot but GAE Classic is still running Go 1.9.

Convert to use GAE Flex so that we can rid ourselves of these limitations (and deploy godoc from the `release-branch.go1.11`)

/cc @bradfitz @dmitshur @katiehockman

**Comment by bradfitz** [comment]
Milestone: Go 1.6  :)

**Comment by bradfitz** [comment]
Also, because we're using a Go 1.10 x/tools on golang.org now, certain links don't work from the release notes & blog post, like:

https://golang.org/cmd/go/#hdr-Modules__module_versions__and_more

Because the godoc heading rules. This should be a heading:

**Modules, module versions, and more**

But it renders as a plain paragraph at the moment.

**Comment by dmitshur** [comment]
Similarly, this statement from https://golang.org/doc/go1.11#godoc is not yet true on https://golang.org/pkg/:

> The `godoc` web server now shows which version of Go introduced new API features.

As a workaround, in the meantime, you can refer to https://tip.golang.org/pkg/. E.g.:

https://tip.golang.org/pkg/os/#UserCacheDir

**Comment by steren** [comment]
FYI, Go 1.11 on the App Engine standard environment is coming.

**Comment by bradfitz** [comment]
@steren, that doesn't help us. App Engine's Go runtime by definition will always be behind what we need, because we have to release Go before App Engine can release a Go runtime for that version. So we can't use App Engine's provided runtimes. We can only do a custom runtime (BYODockerfile).

So we're moving off App Engine classic.

**Comment by steren** [comment]
I see, yes in that case, a Dockerfile approach seems to be appropriate.

**Comment by broady** [comment]
Checklist of things to do.

- [ ] Create Dockerfile
- [ ] Update deployment process/script to use `env: flex`
- [ ] Consider introducing another build flag to include "golang.org" stuff (short, dl). Currently the "appengine" build tag is used for this (see appinit.go). An environment variable might be even better.

Done already: networking/routing. We did this work already about a year ago.

Cannot use "google.golang.org/appengine" packages in Flex, so these must be migrated.

Main things that need to be updated:

- [ ] cmd/godoc/appinit.go (main entrypoint for golang.org)
- [ ] godoc/dl (downloads page uses datastore/memcache. using an in-memory cache for dl is probably fine here)
- [ ] godoc/short (short URL facility, uses datastore/memcache. in-memory cache probably fine here too)

**Comment by broady** [comment]
One more thing we need to do: move the shortlink admin service to another GAE app or just a local command.

It uses `appengine/user` to check the viewer has admin access. There's no simple migration for that (need to set up OAuth/Firebase Auth).

**Comment by bradfitz** [comment]
How about we migrate the shortlinks to a plain text file stored in git?

The admin service is broken even for classic App Engine on custom domains anyway and when I want to add a new link, I spend more time remembering/finding the ugly workaround URL than I would if I just TBR'd a CL to a data file.

We can cache the data file in memory and on cache miss (for new links) fault to Gerrit for fetching the file at HEAD, rate limited to once every 5 minutes or so.

**Comment by broady** [comment]
I like that, it also gives non-admins a list of available shortlinks.

Where should the list live? `x/tools/short/links.tsv`?

**Comment by bradfitz** [comment]
That works. But slight preference for `links.txt`(or links.dat) instead and not requiring tabs. I'd just use any whitespace. Ignore comment lines starting with '#' otherwise just use strings.Fields and get first element is short key and second is full URL.

**Comment by bradfitz** [comment]
It's now been over a month for this "Soon" bug.

What remains?

**Comment by andybons** [comment]
Ping @broady, who last I heard had a working version locally and was working on the deploy scripts.

**Comment by broady** [comment]
Update... what remains:

* porting over the patches in existing deployment scripts (namely, google analytics injection)
* package listing is missing, but package docs work just fine (this is because the package listing looks at GOROOT/pkg, which is empty because I do a clean checkout of the git repo)
* update hostname filtering to allow newly deployed versions - so we can smoke test before migrating traffic to the new version (hostnames look like 20180907t155022-dot-golang-org.appspot.com)

Fast-follow:

* ability to add shortlinks (old app had a login, GAE Flex doesn't have authed handlers)

**Comment by broady** [comment]
Update: well, I sank several hours into figuring out why the package listing was empty.

It looks like there have been numerous changes to [x/tools/godoc/vfs](https://github.com/golang/tools/commits/master/godoc/vfs), my best guess is [CL 101295](https://golang.org/cl/101295).

Precisely, this change to zipfs:
https://github.com/golang/tools/commit/16d1af8d88c08b6a22e07a0900aad5279a7fb4fd#diff-6b09edd71f61887ea342b82eb207149b

/cc @agnivade

**Comment by agnivade** [comment]
Yes, I made that change. But it was also fixed later with https://github.com/golang/go/issues/27162. Are you sure you have the latest 1.11 ?

So this comment - 
> this is because the package listing looks at GOROOT/pkg, which is empty because I do a clean checkout of the git repo

is not true anymore. Package listing will look at the resolved goroot now - which should be the `bin/` folder inside a go distribution where `godoc`, and `go` are kept together. Plus, if there is any confusion, you can pass the `-goroot` flag yourself.

Let me know if I can be of any further help.

**Comment by broady** [comment]
> Yes, I made that change. But it was also fixed later with #27162. Are you sure you have the latest 1.11 ?

It's broken for zipfs. It works fine on real filesystem.

**Comment by agnivade** [comment]
Ah, my apologies for this ! Will send a fix.

**Comment by agnivade** [comment]
Now that site is live, is there anything else pending on this ?

**Comment by broady** [comment]
@agnivade yes, see the unchecked items in the thread.

I'm also going to:

- [ ] build Docker image using Cloud Build, so the only dependency (other than Make) for the person deploying is gcloud. At present, the person deploying must have Docker installed.

**Comment by agnivade** [comment]
Ah thanks. Though I think removing zipfs is orthogonal to this issue, and should not block finishing this task.

**Comment by agnivade** [comment]
@broady - I still don't see the version tags that are supposed to appear.

See - https://tip.golang.org/pkg/net/http/#SameSite vs https://golang.org/pkg/net/http/#SameSite. There is a "1.11" in the previous. Am I missing something ?

**Comment by broady** [comment]
Yes, I was wondering about that, too. I haven't debugged it, yet.

**Comment by dmitshur** [comment]
It seems this issue is mostly resolved, with a minor cleanup followup task left. We should either remove the [Soon](https://github.com/golang/go/labels/Soon) label, since I don't think it applies now, or close this and open a new issue for the remaining cleanup task.

(/cc @andybons since you added the Soon label.)

**Comment by dmitshur** [comment]
The task described in the original issue has long been complete. This issue has remained opened because there was one remaining clean up task (see https://github.com/golang/go/issues/27205#issuecomment-435172401):

> - [ ] remove build tag from main.go (i.e., merge appinit.go and main.go)

The work happening in #29206 obsoletes the need for this to be done. After the golang.org website is deployed from x/website, x/tools/cmd/godoc will be reduced in scope to just serve documentation locally, and its appinit.go will be simply deleted.

That work is better tracked by other open issues, so I'll close this as done.

**Comment by agnivade** [comment]
There were a few others also which wasn't done.

- [ ] remove zipping from build process (generate-index.bash)
- [ ] ability to add shortlinks (old app had a login, GAE Flex doesn't have authed handlers)

Not sure if they would be obsoleted too.
